test/mock_relay.cpp

```diff
@@ -644,17 +718,46 @@ TEST(relay, server_callback) {
   config.state_db = state_db;
   config.local_sock = -1;
 
-  // negative case testing
-  try {
-    server_callback(0, 0, &config);
-  }
-  catch (const std::exception& e) {
-    EXPECT_TRUE(false);
-  }
+   // same name to override static global variable
+  uint8_t server_recv_buffer[BUFFER_SIZE] = {
+    0x0d, 0x00, 0x20, 0x01, 0x0d, 0xb8, 0x01, 0x5a,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x01, 0xfe, 0x80, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x01, 0x00, 0x12, 0x00, 0x03, 0x47, 0x69,
+    0x32, 0x00, 0x09, 0x00, 0x54, 0x07, 0x4f, 0x6d,
+    0x04, 0x00, 0x03, 0x00, 0x28, 0xb0, 0x12, 0xe8,
+    0xb4, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x05, 0x00, 0x18, 0x20, 0x01, 0x0d,
+    0xb8, 0x01, 0x5a, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x01, 0x78, 0x00, 0x00, 0x1c,
+    0x20, 0x00, 0x00, 0x1d, 0x4c, 0x00, 0x01, 0x00,
+    0x0e, 0x00, 0x01, 0x00, 0x01, 0x25, 0x3a, 0x37,
+    0xb9, 0x5a, 0xc6, 0xb0, 0x12, 0xe8, 0xb4, 0x00,
+    0x02, 0x00, 0x0e, 0x00, 0x01, 0x00, 0x01, 0x25,
+    0x3a, 0x32, 0x33, 0x50, 0xe5, 0x49, 0x50, 0x9e,
+    0x40
+  };
+  auto msg_len = sizeof(server_recv_buffer);
+
+  // cover buffer_sz <= 0
+  EXPECT_GLOBAL_CALL(recvfrom, recvfrom(_, _, _, _, _, _)).Times(5).WillOnce(Return(0))
+    .WillOnce(Return(2)).WillOnce(Return(0))
+    .WillOnce(Return(msg_len)).WillOnce(Return(0));
+  ASSERT_NO_THROW(server_callback(0, 0, &config));
+  // cover 0 < buffer_sz < sizeof(struct dhcpv6_msg)
+  ASSERT_NO_THROW(server_callback(0, 0, &config));
+
+  ASSERT_NO_THROW(server_callback(0, 0, &config));
```

Comment:
Why is this assertion called three times?

Reply:
In this case we mock recvfrom 5 times with different return value by EXPECT_GLOBAL_CALL( ).WillOnce(Return()).WillOnce(Return()).WillOnce(Return())

[{0}, {2, 0}, {msg_len, 0}]

Call server_callback() three times will go though this return values to make sure we covered all if else branch in this function.